### ISSUE TITLE
Support for encryption of fields in subclass

### ISSUE THREAD

**Issue by ramasa** [bolcom/spring-data-mongodb-encrypt#8]
```
@Document
class Transaction{
   @Field
   List<BC> list;
}

class BC{
}
class DC extends BC{
   @Field
   @Encrypted
   String field;
}
```

**Comment by agoston** [comment]
Let me explain why this is not supported.

When you do the above scenario, you are relying on the _class field that spring-data-mongo writes to each and every object in mongodb. This has some very serious drawbacks:

- your database becomes tightly coupled to your java code. E.g., you can't just use another piece of code to access the database, or during refactoring java code, you will have to take extra steps to make it backwards compatible. Just simply changing a java class name or package would break this.

- it puts a massive extra data in the database, for no good reason. The full package+class is put there.

- I'm unsure about security, but the fact that spring-mongo-data would happily try loading and instantiateing any class specified in the DB sounds like a gaping security hole.

All in all, I _always_ install a custom mongo mapper that does not put type mapping in the database:

```java
    @Override
    @Bean
    public MappingMongoConverter mappingMongoConverter() throws Exception {
        MappingMongoConverter converter = super.mappingMongoConverter();
        // NB: without overriding defaultMongoTypeMapper, an _class field is put in every document
        // since we know exactly which java class a specific document maps to, this is surplus
        converter.setTypeMapper(new DefaultMongoTypeMapper(null));
        return converter;
    }
```

**tl;dr**: while it's perfectly possible to do what you ask, it is bad design that should not be done in the first place.

**Comment by agoston** [comment]
If we go without ReflectionCache, and do runtime reflection, this is a very easy task to accomplish.

Need to test the performance impact thereof.

**Comment by melbarra88** [comment]
Hi,
you can also take inspiration from Jackson (find an example below) and create a new annotation that allows other projects to explicitly point the subclasses : 

```java
@JsonSubTypes({ 
  @Type(value = Car.class, name = "car"), 
  @Type(value = Truck.class, name = "truck") 
})
public abstract class Vehicle {

}
```

**Comment by TehBakker** [comment]
Hey there,
Great lib, any news on sublcass support ? Kind of a blocking issue for me.

Cheers

**Comment by agoston** [comment]
I've added support for this in 1.3.0-SNAPSHOT. Give it a try, if you like. I still need to update docs before making a release.

**Comment by agoston** [comment]
Released 1.3.0 to maven central.
